File: examples/javascript/httprequest.js
var request = require('request');

var responses = {};
var pointer = 0;

module.exports = function (data, complete) {
	var parts = data.split('|');
	var cmd = parts[0];
	var arg = parts[1] || '';
	pointer++;
	switch (cmd) {
		case 'send':
			var index = pointer;
			request(arg, function (error, response, body) {
				console.log('request complete ' + arg + '. ID: ' + index.toString());
				responses[index] = (error) ? 'error\n' + error : body;
			});
			console.log('sending request ' + arg + '. ID: ' + index.toString());
			complete(null, index.toString());
			break;
		case 'response':
			arg = parseInt(arg);
			if (!responses[arg]) 
				return complete(null, 'wait');
			var response = responses[arg];
			delete responses[arg];
			complete(null, response);
			break;
		default:
			complete(null, 'error\nunknown command');
	};
};

File: examples/cs/httprequest.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Threading.Tasks;

public class Startup
{
    private static Dictionary<int, Task<string>> _tasks = new Dictionary<int, Task<string>>();
    private static int _taskPointer = 0;

    public static string Invoke(string input)
    {
        var parts = input.Split('|');
        var response = string.Empty;
        Task<string> task = null;
        switch (parts[0])
        {
            case "send":
                var pointer = _taskPointer++;
                _tasks.Add(pointer, Task.Factory.StartNew<string>(() => {return Request(parts[1]);}));
                response = pointer.ToString();
                break;
            case "response":
                if (!_tasks.TryGetValue(Convert.ToInt32(parts[1]), out task))
                {
                    response = "error";
                    break;
                }
                if (!task.IsCompleted)
                {
                    response = "wait";
                    break;
                }
                response = task.Result;
                break;
        }
        return response;
    }

    private static string Request(string url)
    {
        string html = string.Empty;
        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
        request.AutomaticDecompression = DecompressionMethods.GZip;
        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        using (Stream stream = response.GetResponseStream())
        using (StreamReader reader = new StreamReader(stream))
        {
            html = reader.ReadToEnd();
        }
        return html;
    }
}